website/content/docs/configuration/server.mdx

```diff
@@ -135,15 +135,33 @@ server {
   a tradeoff as it lowers failure detection time of nodes at the tradeoff of
   false positives and increased load on the leader.
 
-- `failover_heartbeat_ttl` `(string: "5m")` - Specifies the TTL applied to
-	heartbeats after a new leader is elected, since we no longer know the status
-	of all the heartbeats. This is specified using a label suffix like "30s" or
-	"1h".
+- `failover_heartbeat_ttl` `(string: "5m")` - Specifies the initial TTL applied to
+  Client heartbeats after a new leader is elected. This TTL must be high enough
+  such that a Client previously connected to a now-unresponsive server will be
+  able to find a new healthy server through which to send heartbeats to the new
+  leader. The strict minimum value depends on the number of nodes in the cluster,
+  the value of `max_heartbeats_per_second`, `min_heartbeat_ttl`, and `heartbeat_grace`.
+  As the number of nodes in the cluster increases, the longer the TTL each client
+  is assigned. This value is uniformly randomized between the values computed by
+  these two formula:
 
-  ~> Lowering the `failover_heartbeat_ttl` is a tradeoff as it lowers failure
+```
+<min ttl> = max(((<num nodes> / <max_heartbeats_per_second>) + <heartbeat_grace>), <min_heartbeat_ttl>)
+<max ttl> = max(((2 * (<num nodes> / <max_heartbeats_per_second>)) + <heartbeat_grace>), <min_heartbeat_ttl>)
+```
+
+  The default value of `"5m"` is fine for small and medium sized clusters of fewer
+  than approximately 5k nodes - the value of `failover_heartbeat_ttl` must be
+  higher than `<max ttl>`, ideally with some extra padding for clusters growing
+  over time. If `<max ttl>` exceeds the failover heartbeat ttl, the cluster may
```

Comment:
Seems unfortunate to use a lot of approximation right after giving a formula that seems to guide toward the right answer.

For example a 5,000 node cluster with everything else on defaults gets you:

```
min -> (5000 / 50) + 10 = 110 seconds (1m50s)
max -> (2 * (5000 / 50)) + 10 = 210 seconds (3m30s)
```

So on paper it would seem `failover_heartbeat_ttl = "5m"` would be ample time for all Clients to heartbeat before being considered `down` by the new leader. If we're suggesting otherwise I'm curious why.

Given the scenario:

1. `t=1s` Client C1 heartbeats to Server S1 who is the Leader, is informed to heartbeat again within 1m40s (min - grace)
2. `t=2s` Client C2 heartbeats to Server S1, is informed to heartbeat again within 3m20s (max - grace)
3. `t=10s` Server S1 crashes
4. `t=20s` Server S2 elected Leader, sets all clients ttls to`t=5m20s` (wallclock + `5m`)
5. `t=1m1s` Client C1 attempts to heartbeat to dead server...
6. `t=3m22s` Client C2 attempts to heartbeat to dead server...
7. ??? ...118 seconds pass... ???
8. `t=5m20s` Server S2 marks Clients C1 & C2 as `down` and allocations as `lost`

In the absence of all servers known to the clients being down and a Consul failure preventing finding new servers, I don't know how to account for a Client not being able to find a live server within `3m35s`. That being said I can't find where we set TCP timeouts (other than a dial timeout of 10s which wouldn't apply to existing connections), so if the clients do not receive an `connection reset by peer` error (they should) perhaps they block indefinitely? There's a 5s RPC timeout that only seems to affect how retries are performed *after* the attempted RPC fails... so if the RPC's write simply blocks indefinitely... perhaps this could occur?

Reply:
> Seems unfortunate to use a lot of approximation right after giving a formula that seems to guide toward the right answer.

So long as we're noting that we're giving room for growth (but also dynamic node scaling which will make the count less stable), I think this sentence says to users "if you keep the defaults and you're below roughly this size, don't worry about this", which I think is good messaging to when you should _start_ thinking about it as a cluster admin.

Reply:
> 7. ??? ...118 seconds pass... ???

For completeness: I found a Yamux timeout that could theoretically add **150s** of latency here (since the timeout is hit twice if a client fails over *and* the server it connects to fails over as well): https://github.com/hashicorp/yamux/blob/0bc27b27de87381b58e35462a8dcbd69040279c1/mux.go#L66

Now whether that timeout *should* be that or not is a question for a different issue/PR...